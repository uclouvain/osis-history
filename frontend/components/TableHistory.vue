<!--
  -
  -   OSIS stands for Open Student Information System. It's an application
  -   designed to manage the core business of higher education institutions,
  -   such as universities, faculties, institutes and professional schools.
  -   The core business involves the administration of students, teachers,
  -   courses, programs and so on.
  -
  -   Copyright (C) 2015-2021 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
  -
  -   This program is free software: you can redistribute it and/or modify
  -   it under the terms of the GNU General Public License as published by
  -   the Free Software Foundation, either version 3 of the License, or
  -   (at your option) any later version.
  -
  -   This program is distributed in the hope that it will be useful,
  -   but WITHOUT ANY WARRANTY; without even the implied warranty of
  -   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  -   GNU General Public License for more details.
  -
  -   A copy of this license - GNU General Public License - is available
  -   at the root of the source code of this program.  If not,
  -   see http://www.gnu.org/licenses/.
  -
  -->
<template>
  <table class="table table-striped">
    <thead v-if="!onHeadersRender">
      <tr>
        <th>{{ $t('date') }}</th>
        <th>{{ $t('message') }}</th>
        <th>{{ $t('author') }}</th>
      </tr>
    </thead>
    <thead
        v-else
        v-html="renderedHeaders"
    />

    <tbody v-if="entries.length && !onItemRender">
      <tr
          v-for="(entry, index) in entries"
          :key="index"
      >
        <td>{{ entry.created }}</td>
        <td :inner-html.prop="filterXssAndFormat(entry.message)" />
        <td>{{ entry.author }}</td>
      </tr>
    </tbody>
    <tbody
        v-else-if="entries.length && !!onItemRender"
        v-html="renderedRows"
    />
    <tbody
        v-else-if="!!onHistoryEmptyRender"
        v-html="renderedEmpty"
    />
    <tbody v-else>
      <tr>
        <td colspan="3">
          {{ $t('no_entry') }}
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script lang="ts">
import {defineComponent} from "vue";
import type {PropType} from "vue";
import type {Entry} from "../interfaces";
import {filterXssAndFormat} from "../utils";

export default defineComponent({
  name: 'TableHistory',
  props: {
    entries: {
      type: Array as PropType<Entry[]>,
      required: true,
    },
    onItemRender: {
      type: Function as PropType<(entry: Entry) => string>,
      default: null,
    },
    onHeadersRender: {
      type: Function as PropType<(entries: Entry[]) => string>,
      default: null,
    },
    onHistoryEmptyRender: {
      type: Function as PropType<(mode: string) => string>,
      default: null,
    },
  },
  data: function () {
    return {
      // Render dynamically the table headers and items
      renderedHeaders: this.onHeadersRender ? this.onHeadersRender(this.entries) : '',
      renderedRows: this.onItemRender ? this.entries.map(this.onItemRender).join('') : [],
      renderedEmpty: this.onHistoryEmptyRender ? this.onHistoryEmptyRender('table') : '',
    };
  },
  methods: {filterXssAndFormat},
});
</script>

<style>
.history-viewer .table {
  background-color: #fff;
}
</style>
