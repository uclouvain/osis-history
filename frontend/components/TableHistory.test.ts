/*
 *
 * OSIS stands for Open Student Information System. It's an application
 * designed to manage the core business of higher education institutions,
 * such as universities, faculties, institutes and professional schools.
 * The core business involves the administration of students, teachers,
 * courses, programs and so on.
 *
 * Copyright (C) 2015-2023 Universit√© catholique de Louvain (http://www.uclouvain.be)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of this license - GNU General Public License - is available
 * at the root of the source code of this program.  If not,
 * see http://www.gnu.org/licenses/.
 *
 */

import {expect, test} from 'vitest';
import {mount} from '@vue/test-utils';
import type {Entry} from "../interfaces";
import TableHistory from "./TableHistory.vue";

const mockEntries: Entry[] = [
  {
    author: 'John Doe',
    created: '06/04/2021 16:23',
    message: 'Ut molestiae aliquam tenetur est veritatis dolorum tempora. Eaque similique et qui ea consequatur illum dignissimos. Ullam qui vel qui quo. Ea dolor quasi quibusdam quia illum sunt aspernatur. Eveniet odio quos ad.',
  }, {
    author: 'Alice Smith',
    created: '03/04/2021 16:23',
    message: 'Nulla natus dolores dolor. Voluptatem numquam doloribus architecto non praesentium.\nSit facilis vitae sapiente. Placeat est qui et.',
  }, {
    author: 'Bob Martin',
    created: '01/04/2021 08:23',
    message: 'Corrupti <strong>reiciendis</strong> laboriosam repellat adipisci quis.\n\nIure omnis cum nihil accusantium ut doloribus nisi. Ut itaque suscipit et nulla iste enim. Assumenda quis et ullam temporibus quidem rerum nostrum. Rerum debitis quae qui ea molestiae animi minus optio.',
  }, {
    author: 'John Doe',
    created: '28/03/2021 10:00',
    message: '',
  },
];


test('empty history', () => {
  const wrapper = mount(TableHistory, {
    props: {entries: []},
  });
  expect(wrapper.html()).toMatchSnapshot();
});

test('vertical history', () => {
  const wrapper = mount(TableHistory, {
    props: {entries: mockEntries},
  });
  expect(wrapper.html()).toMatchSnapshot();
});

test('horizontal history', () => {
  const wrapper = mount(TableHistory, {
    props: {entries: mockEntries, horizontal: true},
  });
  expect(wrapper.html()).toMatchSnapshot();
});

test('custom rendering with entries', () => {
  const wrapper = mount(TableHistory, {
    props: {
      entries: mockEntries,
      onHeadersRender: () => (`<tr><th>Foo</th></tr>`),
      onItemRender: (item: Entry) => (`<tr><td>${item.message}</td></tr>`),
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});

test('custom rendering empty', () => {
  const wrapper = mount(TableHistory, {
    props: {
      entries: [],
      onHistoryEmptyRender: () => (`<tr colspan="3"><td>Nope, nothing to display</td></tr>`),
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});
