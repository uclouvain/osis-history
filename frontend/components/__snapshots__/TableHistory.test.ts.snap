// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom rendering empty 1`] = `
"<table class=\\"table table-striped\\">
  <thead>
    <tr>
      <th>date</th>
      <th>message</th>
      <th>author</th>
    </tr>
  </thead>
  <tbody>
    <tr colspan=\\"3\\">
      <td>Nope, nothing to display</td>
    </tr>
  </tbody>
</table>"
`;

exports[`custom rendering with entries 1`] = `
"<table class=\\"table table-striped\\">
  <thead>
    <tr>
      <th>Foo</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Ut molestiae aliquam tenetur est veritatis dolorum tempora. Eaque similique et qui ea consequatur illum dignissimos. Ullam qui vel qui quo. Ea dolor quasi quibusdam quia illum sunt aspernatur. Eveniet odio quos ad.</td>
    </tr>
    <tr>
      <td>Nulla natus dolores dolor. Voluptatem numquam doloribus architecto non praesentium.
        Sit facilis vitae sapiente. Placeat est qui et.</td>
    </tr>
    <tr>
      <td>Corrupti <strong>reiciendis</strong> laboriosam repellat adipisci quis.

        Iure omnis cum nihil accusantium ut doloribus nisi. Ut itaque suscipit et nulla iste enim. Assumenda quis et ullam temporibus quidem rerum nostrum. Rerum debitis quae qui ea molestiae animi minus optio.</td>
    </tr>
    <tr>
      <td></td>
    </tr>
  </tbody>
</table>"
`;

exports[`empty history 1`] = `
"<table class=\\"table table-striped\\">
  <thead>
    <tr>
      <th>date</th>
      <th>message</th>
      <th>author</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td colspan=\\"3\\">no_entry</td>
    </tr>
  </tbody>
</table>"
`;

exports[`horizontal history 1`] = `
"<table class=\\"table table-striped\\" horizontal=\\"true\\">
  <thead>
    <tr>
      <th>date</th>
      <th>message</th>
      <th>author</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>06/04/2021 16:23</td>
      <td>
        <p>Ut molestiae aliquam tenetur est veritatis dolorum tempora. Eaque similique et qui ea consequatur illum dignissimos. Ullam qui vel qui quo. Ea dolor quasi quibusdam quia illum sunt aspernatur. Eveniet odio quos ad.</p>
      </td>
      <td>John Doe</td>
    </tr>
    <tr>
      <td>03/04/2021 16:23</td>
      <td>
        <p>Nulla natus dolores dolor. Voluptatem numquam doloribus architecto non praesentium.<br>Sit facilis vitae sapiente. Placeat est qui et.</p>
      </td>
      <td>Alice Smith</td>
    </tr>
    <tr>
      <td>01/04/2021 08:23</td>
      <td>
        <p>Corrupti &lt;strong&gt;reiciendis&lt;/strong&gt; laboriosam repellat adipisci quis.</p>

        <p>Iure omnis cum nihil accusantium ut doloribus nisi. Ut itaque suscipit et nulla iste enim. Assumenda quis et ullam temporibus quidem rerum nostrum. Rerum debitis quae qui ea molestiae animi minus optio.</p>
      </td>
      <td>Bob Martin</td>
    </tr>
    <tr>
      <td>28/03/2021 10:00</td>
      <td></td>
      <td>John Doe</td>
    </tr>
  </tbody>
</table>"
`;

exports[`vertical history 1`] = `
"<table class=\\"table table-striped\\">
  <thead>
    <tr>
      <th>date</th>
      <th>message</th>
      <th>author</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>06/04/2021 16:23</td>
      <td>
        <p>Ut molestiae aliquam tenetur est veritatis dolorum tempora. Eaque similique et qui ea consequatur illum dignissimos. Ullam qui vel qui quo. Ea dolor quasi quibusdam quia illum sunt aspernatur. Eveniet odio quos ad.</p>
      </td>
      <td>John Doe</td>
    </tr>
    <tr>
      <td>03/04/2021 16:23</td>
      <td>
        <p>Nulla natus dolores dolor. Voluptatem numquam doloribus architecto non praesentium.<br>Sit facilis vitae sapiente. Placeat est qui et.</p>
      </td>
      <td>Alice Smith</td>
    </tr>
    <tr>
      <td>01/04/2021 08:23</td>
      <td>
        <p>Corrupti &lt;strong&gt;reiciendis&lt;/strong&gt; laboriosam repellat adipisci quis.</p>

        <p>Iure omnis cum nihil accusantium ut doloribus nisi. Ut itaque suscipit et nulla iste enim. Assumenda quis et ullam temporibus quidem rerum nostrum. Rerum debitis quae qui ea molestiae animi minus optio.</p>
      </td>
      <td>Bob Martin</td>
    </tr>
    <tr>
      <td>28/03/2021 10:00</td>
      <td></td>
      <td>John Doe</td>
    </tr>
  </tbody>
</table>"
`;
